lib/mlir/Dialect/Clift/IR/ImportModel.cpp

```diff
@@ -0,0 +1,599 @@
+//
+// Copyright (c) rev.ng Labs Srl. See LICENSE.md for details.
+//
+
+#include "llvm/ADT/SmallSet.h"
+#include "llvm/Support/FormatVariadic.h"
+
+#include "revng/ADT/RecursiveCoroutine.h"
+
+#include "revng-c/mlir/Dialect/Clift/IR/Clift.h"
+#include "revng-c/mlir/Dialect/Clift/IR/CliftAttributes.h"
+#include "revng-c/mlir/Dialect/Clift/IR/CliftTypes.h"
+#include "revng-c/mlir/Dialect/Clift/IR/ImportModel.h"
+
+namespace {
+
+namespace clift = mlir::clift;
+
+template<typename Attribute>
+using AttributeVector = llvm::SmallVector<Attribute, 16>;
+
+class CliftConverter {
+  mlir::MLIRContext *Context;
+  llvm::function_ref<mlir::InFlightDiagnostic()> EmitError;
+
+  llvm::DenseMap<uint64_t, clift::TypeDefinition> Cache;
+  llvm::DenseMap<uint64_t, const model::Type *> IncompleteTypes;
+
+  llvm::SmallSet<uint64_t, 16> DefinitionGuardSet;
+
+  class DefinitionGuard {
+    CliftConverter *Self = nullptr;
+    uint64_t ID;
+
+  public:
+    explicit DefinitionGuard(CliftConverter &Self, const uint64_t ID) {
+      if (Self.DefinitionGuardSet.insert(ID).second) {
+        this->Self = &Self;
+        this->ID = ID;
+      }
+    }
+
+    DefinitionGuard(const DefinitionGuard &) = delete;
+    DefinitionGuard &operator=(const DefinitionGuard &) = delete;
+
+    ~DefinitionGuard() {
+      if (Self != nullptr) {
+        Self->DefinitionGuardSet.erase(ID);
+      }
+    }
+
+    explicit operator bool() const { return Self != nullptr; }
+  };
+
+public:
+  explicit CliftConverter(mlir::MLIRContext &Context,
+                          llvm::function_ref<mlir::InFlightDiagnostic()>
+                            EmitError = nullptr) :
+    Context(&Context), EmitError(EmitError) {}
+
+  CliftConverter(const CliftConverter &) = delete;
+  CliftConverter &operator=(const CliftConverter &) = delete;
+
+  ~CliftConverter() {
+    revng_assert(DefinitionGuardSet.empty());
+    revng_assert(IncompleteTypes.empty());
+  }
+
+  clift::ValueType convertType(const model::Type &ModelType) {
+    const clift::ValueType T = getUnwrappedType(ModelType,
+                                                /*RequireComplete=*/true);
+    if (T and not processIncompleteTypes()) {
+      IncompleteTypes.clear();
+      return nullptr;
+    }
+    return T;
+  }
+
+  clift::ValueType convertQualifiedType(const model::QualifiedType &ModelType) {
+    const clift::ValueType T = getQualifiedType(ModelType,
+                                                /*RequireComplete=*/true);
+    if (T and not processIncompleteTypes()) {
+      IncompleteTypes.clear();
+      return nullptr;
+    }
+    return T;
+  }
+
+private:
+  mlir::BoolAttr getBool(bool const Value) {
+    return mlir::BoolAttr::get(Context, Value);
+  }
+
+  mlir::BoolAttr getFalse() { return getBool(false); }
+
+  template<typename T, typename... ArgTypes>
```

Comment:
Can't we use a `concept` to put some constraints on `T`.
The only valid `T`s are `clift::ValueType`s right?

Reply:
It's any MLIR type with static `verify` and `get` functions. It's used for both types and attributes. Maybe I could constrain it somehow, but that doesn't seem particularly useful since it's a very local function.